File: Week16/거리두기확인하기.py
def solution(places):
    answer = []
    row = [[1,1,0],[1,1,0],[-1,-1,0],[-1,-1,0]]
    col = [[0,1,1],[0,-1,-1],[0,1,1],[0,-1,-1]]
    for i in places: # 대기실 개수
        isViolate = False
        for j in range(5): # 하나의 대기실
            for l in range(5): 
                if i[j][l] == 'P': # 응시자일 경우만 사방 체크
                    for k in range(4):
                        hasP = False
                        allX = True
                        for o in range(3):
                            m = j+row[k][o]
                            n = l+col[k][o]
                            if -1<m<5 and -1<n<5:
                                if i[m][n] == 'P':
                                    hasP = True
                                    if abs(row[k][o]) + abs(col[k][o]) < 2:
                                        isViolate = True
                                        break
                                elif i[m][n] == 'O':
                                    allX = False
                                    if abs(row[k][o]) + abs(col[k][o]) == 1:
                                        if abs(row[k][o]) == 1 and -1<m+row[k][o]<5 and i[m+row[k][o]][n] == 'P':
                                            isViolate = True
                                            break
                                        elif abs(col[k][o]) == 1 and -1<n+col[k][o]<5 and i[m][n+col[k][o]] == 'P':
                                            isViolate = True
                                            break
                        if hasP and not allX:  # 지금 조건의 모순은 [PX],[PX] 는 통과시켜버림
                            isViolate = True
                            break
                if isViolate:
                    break
            if isViolate:
                break
        if isViolate:
            answer.append(0)
        else:
            answer.append(1)
                
    return answer

File: Week16/거리두기확인하기.java
import java.util.*;

class Solution {
    
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    
    
    public int[] solution(String[][] places) {
        int[] answer = new int[places.length];
        for(int k=0; k<places.length; k++){
            boolean flag = false;
            for(int i = 0; i<5; i++){
                for(int j = 0; j<5; j++){
                    if(places[k][i].charAt(j) == 'P'){
                        boolean[][] visited = new boolean[5][5];
                        Queue<int[]> queue = new ArrayDeque<>();
                        queue.add(new int[] {i, j, 0});
                        while(!(queue.isEmpty()) && flag == false){
                            int[] cur = queue.poll();
                            if(cur[2] == 2){
                                continue;
                            }
                            visited[cur[0]][cur[1]] = true;
                            for(int d = 0; d<4; d++){
                                int next_x = cur[0] + dx[d];
                                int next_y = cur[1] + dy[d];
                                if(next_x >= 0 && next_x < 5 && next_y>= 0 && next_y < 5 && visited[next_x][next_y] == false){
                                    if(places[k][next_x].charAt(next_y) == 'O'){
                                        queue.add(new int[] {next_x, next_y, cur[2] + 1});
                                    }
                                    else if(places[k][next_x].charAt(next_y) == 'P'){
                                        flag = true;
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        
        if(flag == false){
            answer[k] = 1;
        }else{
            answer[k] = 0;
        }
        }
        return answer;
    }
}